File: java/0242-valid-anagram.java
/**
 * LeetCode 242: Valid Anagram
 *
 * Problem:
 * Given two strings s and t, return true if t is an anagram of s, and false otherwise.
 * An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase,
 * typically using all the original letters exactly once.
 *
 * Example:
 * s = "anagram", t = "nagaram" => true
 * s = "rat", t = "car" => false
 */

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class ValidAnagram {

    // Approach 1: Sorting
    public static boolean isAnagram_Sorting(String s, String t) {
        /**
         * Convert strings to char arrays, sort them, and compare
         * Time: O(n log n), Space: O(1) - not counting output
         */
        char[] sArray = s.toCharArray();
        char[] tArray = t.toCharArray();
        Arrays.sort(sArray);
        Arrays.sort(tArray);
        return Arrays.equals(sArray, tArray);
    }

    // Approach 2: Hash Map
    public static boolean isAnagram_HashMap(String s, String t) {
        /**
         * Count characters in s, then subtract from t
         * Time: O(n), Space: O(k) where k = alphabet size
         */
        if (s.length() != t.length()) return false;

        Map<Character, Integer> charCount = new HashMap<>();

        for (char c : s.toCharArray()) {
            charCount.put(c, charCount.getOrDefault(c, 0) + 1);
        }

        for (char c : t.toCharArray()) {
            if (!charCount.containsKey(c)) return false;
            charCount.put(c, charCount.get(c) - 1);
            if (charCount.get(c) < 0) return false;
        }

        return true;
    }

    // Approach 3: Array (Fixed size for lowercase English)
    public static boolean isAnagram_Array(String s, String t) {
        /**
         * Use array of size 26 for lowercase letters
         * Time: O(n), Space: O(1) - fixed size array
         */
        if (s.length() != t.length()) return false;

        int[] count = new int[26];

        for (int i = 0; i < s.length(); i++) {
            count[s.charAt(i) - 'a']++;
            count[t.charAt(i) - 'a']--;
        }

        for (int c : count) {
            if (c != 0) return false;
        }

        return true;
    }

    // Approach 4: Two Hash Maps
    public static boolean isAnagram_TwoMaps(String s, String t) {
        /**
         * Create frequency maps for both strings and compare
         * Time: O(n), Space: O(k)
         */
        if (s.length() != t.length()) return false;

        Map<Character, Integer> countS = new HashMap<>();
        Map<Character, Integer> countT = new HashMap<>();

        for (char c : s.toCharArray()) {
            countS.put(c, countS.getOrDefault(c, 0) + 1);
        }

        for (char c : t.toCharArray()) {
            countT.put(c, countT.getOrDefault(c, 0) + 1);
        }

        return countS.equals(countT);
    }

    // Test Cases
    public static void main(String[] args) {
        String[][] testCases = {
            {"anagram", "nagaram"},
            {"rat", "car"},
            {"a", "b"},
            {"ab", "ba"},
            {"listen", "silent"},
            {"hello", "world"}
        };

        boolean[] expectedResults = {true, false, false, true, true, false};

        for (int i = 0; i < testCases.length; i++) {
            String s = testCases[i][0];
            String t = testCases[i][1];
            boolean expected = expectedResults[i];

            assert isAnagram_Sorting(s, t) == expected : "Sorting failed";
            assert isAnagram_HashMap(s, t) == expected : "HashMap failed";
            assert isAnagram_Array(s, t) == expected : "Array failed";
            assert isAnagram_TwoMaps(s, t) == expected : "TwoMaps failed";

            System.out.println("s=\"" + s + "\", t=\"" + t + "\" => " + expected + " ✓");
        }

        System.out.println("\nAll tests passed!");

        // Example usage
        System.out.println("\nValid Anagram Example: " + isAnagram_Array("anagram", "nagaram"));
    }
}

File: python/0242-valid-anagram.py
"""
LeetCode 242: Valid Anagram

Problem:
Given two strings s and t, return true if t is an anagram of s, and false otherwise.
An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase,
typically using all the original letters exactly once.

Example:
s = "anagram", t = "nagaram" => True
s = "rat", t = "car" => False
s = "a", t = "b" => False

Time Complexity: O(n log n) for sorting, O(n) for hash map
Space Complexity: O(1) for sorting (if we don't count output), O(k) for hash map (k = alphabet size)
"""

from collections import Counter


# Approach 1: Sorting
def isAnagram_Sorting(s: str, t: str) -> bool:
    """
    Two strings are anagrams if they have the same characters when sorted
    Time: O(n log n), Space: O(1) - not counting the space used by sorting
    """
    return sorted(s) == sorted(t)


# Approach 2: Hash Map / Dictionary
def isAnagram_HashMap(s: str, t: str) -> bool:
    """
    Count frequency of each character in both strings
    If frequencies match, they are anagrams
    Time: O(n), Space: O(k) where k is the alphabet size (26 for lowercase English)
    """
    if len(s) != len(t):
        return False

    char_count = {}

    # Count characters in s
    for char in s:
        char_count[char] = char_count.get(char, 0) + 1

    # Subtract count from t
    for char in t:
        if char not in char_count:
            return False
        char_count[char] -= 1
        if char_count[char] < 0:
            return False

    return True


# Approach 3: Counter from Collections
def isAnagram_Counter(s: str, t: str) -> bool:
    """
    Python's Counter is an elegant way to solve this
    Time: O(n), Space: O(k)
    """
    return Counter(s) == Counter(t)


# Approach 4: Two Hash Maps
def isAnagram_TwoMaps(s: str, t: str) -> bool:
    """
    Create frequency maps for both strings and compare
    """
    if len(s) != len(t):
        return False

    count_s = {}
    count_t = {}

    for i in range(len(s)):
        count_s[s[i]] = count_s.get(s[i], 0) + 1
        count_t[t[i]] = count_t.get(t[i], 0) + 1

    return count_s == count_t


# Approach 5: Array (Fixed size for lowercase English)
def isAnagram_Array(s: str, t: str) -> bool:
    """
    Use an array of size 26 for lowercase English letters
    Time: O(n), Space: O(1) - fixed size array
    """
    if len(s) != len(t):
        return False

    count = [0] * 26

    for i in range(len(s)):
        count[ord(s[i]) - ord("a")] += 1
        count[ord(t[i]) - ord("a")] -= 1

    return all(c == 0 for c in count)


# Test Cases
def test_isAnagram():
    test_cases = [
        ("anagram", "nagaram", True),
        ("rat", "car", False),
        ("a", "b", False),
        ("ab", "ba", True),
        ("abc", "bca", True),
        ("", "", True),
        ("a", "a", True),
        ("listen", "silent", True),
        ("hello", "world", False),
    ]

    for s, t, expected in test_cases:
        assert isAnagram_Sorting(s, t) == expected, f"Sorting failed for {s}, {t}"
        assert isAnagram_HashMap(s, t) == expected, f"HashMap failed for {s}, {t}"
        assert isAnagram_Counter(s, t) == expected, f"Counter failed for {s}, {t}"
        assert isAnagram_TwoMaps(s, t) == expected, f"TwoMaps failed for {s}, {t}"
        assert isAnagram_Array(s, t) == expected, f"Array failed for {s}, {t}"

        print(f's="{s}", t="{t}" => {expected} ✓')


if __name__ == "__main__":
    test_isAnagram()
    print("\nAll tests passed!")

    # Example usage
    print(f'\nValid Anagram Example: {isAnagram_Counter("anagram", "nagaram")}')
